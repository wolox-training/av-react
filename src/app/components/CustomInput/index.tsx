import React, { Ref } from 'react';

import styles from './styles.module.scss';

interface Props extends React.HTMLProps<HTMLInputElement> {
  inputRef?: Ref<HTMLInputElement>;
  error?: string;
  type?: 'text' | 'password' | 'email' | 'number' | 'date';
  inputStyle: string;
  labelText?: string;
  labelStyle?: string;
}

export default function CustomInput({
  inputRef,
  error,
  type = 'text',
  inputStyle,
  labelText,
  labelStyle,
  ...props
}: Props) {
  return (
    <div className={`${styles.wrapper} column`}>
      {labelText && <label className={labelStyle}>{labelText}</label>}
      <input {...props} ref={inputRef} type={type} className={inputStyle} />
      {error && <span className={styles.errorLabels}>{error}</span>}
    </div>
  );
}
