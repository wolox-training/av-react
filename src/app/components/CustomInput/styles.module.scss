@import 'variables/colors';

.input-container {
  margin: 16px 25px 0;
}

.custom-input {
  border-radius: 5px;
  height: 32px;
  margin-top: 8px;
  padding-left: 5px;
  width: 100%;
}

.error-labels {
  color: $red;
}
