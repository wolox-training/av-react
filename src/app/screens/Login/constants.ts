export const LOGIN_FIELDS = {
  email: 'email',
  password: 'password'
};
