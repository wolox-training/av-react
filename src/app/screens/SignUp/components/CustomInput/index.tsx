import React from 'react';

import styles from './styles.module.scss';

interface Props extends React.HTMLProps<HTMLInputElement> {
  inputRef?: (ref: HTMLInputElement) => void;
  error?: string;
}

export default function CustomInput({ inputRef, error, ...props }: Props) {
  return (
    <>
      <input {...props} ref={inputRef} className={styles.customInputs} />
      {error && <span className={styles.errorLabels}>{error}</span>}
    </>
  );
}
