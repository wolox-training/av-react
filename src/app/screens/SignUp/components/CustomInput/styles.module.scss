@import '../../../../../scss/variables/colors';

.custom-inputs {
  border-radius: 5px;
  height: 32px;
  margin: 8px 23px 0 25px;
}

.error-labels {
  color: $red;
  margin-left: 25px;
}
