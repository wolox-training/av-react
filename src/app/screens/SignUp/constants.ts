export const SIGN_UP_FIELDS = {
  firstName: 'firstName',
  lastName: 'lastName',
  email: 'email',
  password: 'password',
  passwordConfirmation: 'passwordConfirmation'
};
