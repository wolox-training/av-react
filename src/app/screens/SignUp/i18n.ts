import i18next from 'i18next';

i18next.addResources('es', 'SignUp', {
  logoAlt: 'WoloxLogo',
  firstName: 'Nombre',
  lastName: 'Apellido',
  email: 'Email',
  password: 'Password',
  signUp: 'Sign up',
  login: 'Login',
  confirmPassword: 'Confirmación de Password',
  required: 'Requerido',
  invalidEmail: 'Email inválido',
  passwordsNoMatch: 'Los passwords no coinciden'
});
