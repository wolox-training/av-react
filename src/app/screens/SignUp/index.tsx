import React from 'react';
import i18next from 'i18next';
import { useForm } from 'react-hook-form';
import { Link } from 'react-router-dom';

import { User } from '~utils/types';
import { signUp } from '~services/UserService';
import CustomErrorDisplayer from '~components/CustomErrorDisplayer';
import Loading from '~components/Spinner/components/loading';
import CustomInput from '~components/CustomInput';
import { PATHS } from '~constants/paths';
// eslint-disable-next-line import/namespace
import { useLazyRequest } from '~hooks/useRequest';

import WoloxImg from '../Assets/wolox-logo.png';

import { SIGN_UP_FIELDS } from './constants';
import styles from './styles.module.scss';

interface FormData extends User {
  passwordConfirmation?: string;
}

export default function SignUp() {
  const { register, handleSubmit, watch, errors } = useForm<FormData>({ mode: 'all' });
  const [, loading, error, sendRequest] = useLazyRequest({
    request: signUp
  });
  const onSubmit = handleSubmit(data => {
    data.locale = i18next.language;
    sendRequest(data);
  });

  return (
    <div className="column center">
      <form onSubmit={onSubmit} className={`column ${styles.signupForm}`}>
        <img src={WoloxImg} alt={i18next.t('SignUp:logoAlt') as string} className={styles.logo} />
        <CustomInput
          name={SIGN_UP_FIELDS.firstName}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.firstName?.message}
          label={i18next.t('SignUp:firstName') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.lastName}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.lastName?.message}
          label={i18next.t('SignUp:lastName') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.email}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') },
            pattern: {
              value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i,
              message: i18next.t('SignUp:invalidEmail')
            }
          })}
          error={errors.email?.message}
          type="email"
          label={i18next.t('SignUp:email') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.password}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') }
          })}
          error={errors.password?.message}
          type="password"
          label={i18next.t('SignUp:password') as string}
          labelClassName={styles.customLabel}
        />
        <CustomInput
          name={SIGN_UP_FIELDS.passwordConfirmation}
          inputRef={register({
            required: { value: true, message: i18next.t('SignUp:required') },
            validate: value =>
              value === watch(SIGN_UP_FIELDS.password) || (i18next.t('SignUp:passwordsNoMatch') as string)
          })}
          error={errors.passwordConfirmation?.message}
          type="password"
          label={i18next.t('SignUp:confirmPassword') as string}
          labelClassName={styles.customLabel}
        />
        <div className={`column ${styles.signupButtonContainer}`}>
          <button
            className={`${styles.customButton} ${styles.signupButton} `}
            type="submit"
            disabled={loading}
          >
            {i18next.t('SignUp:signUp')}
          </button>
        </div>
        <Link to={PATHS.login} className="row">
          <button
            className={`${styles.customButton} ${styles.loginButton}  full-width`}
            type="button"
            disabled={loading}
          >
            {i18next.t('SignUp:login')}
          </button>
        </Link>
      </form>
      {loading && <Loading className={styles.loading} />}
      {error?.errorData?.errors.fullMessages.length && (
        <CustomErrorDisplayer errors={error.errorData?.errors.fullMessages} />
      )}
    </div>
  );
}
