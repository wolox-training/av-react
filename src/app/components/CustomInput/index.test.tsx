import React from 'react';
import { render, screen } from '@testing-library/react';

import { SIGN_UP_FIELDS } from '../../screens/SignUp/constants';

import CustomInput from './';

const register = jest.fn();
const ERROR_MESSAGE = 'This is a error message';
const MINIMUM_ALERT_LENGTH = 1;

describe('Custom input test', () => {
  test('Should match snapshot', () => {
    const { container } = render(<CustomInput />);
    expect(container).toMatchSnapshot();
  });

  test('Should render input type text as default', () => {
    render(
      <CustomInput name={SIGN_UP_FIELDS.firstName} inputRef={register} error="" label="SignUp:firstName" />
    );
    expect(screen.getByLabelText('SignUp:firstName').getAttribute('type')).toBe('text');
  });

  test('Should render error message', () => {
    render(
      <CustomInput
        name={SIGN_UP_FIELDS.firstName}
        inputRef={register}
        error={ERROR_MESSAGE}
        label="SignUp:firstName"
      />
    );
    expect(screen.getAllByRole('alert').length).toBe(MINIMUM_ALERT_LENGTH);
  });

  test('Should render label text', () => {
    render(
      <CustomInput name={SIGN_UP_FIELDS.firstName} inputRef={register} error="" label="SignUp:firstName" />
    );
    expect(screen.getByText('SignUp:firstName')).toBeInTheDocument();
  });
});
