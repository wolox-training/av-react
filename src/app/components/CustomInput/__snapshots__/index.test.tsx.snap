// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom input test Should match snapshot 1`] = `
<div>
  <div
    class="inputContainer column"
  >
    <input
      class="input full-width"
      type="text"
    />
  </div>
</div>
`;
